nipype/utils/config.py

```diff
@@ -242,4 +252,77 @@ def resource_monitor(self, value):
                                      ('%s' % self._resource_monitor).lower())
 
     def enable_resource_monitor(self):
+        """Sets the resource monitor on"""
         self.resource_monitor = True
+
+    def get_display(self):
+        """Returns the first display available"""
+
+        # Check if an Xorg server is listening
+        # import subprocess as sp
+        # if not hasattr(sp, 'DEVNULL'):
+        #     setattr(sp, 'DEVNULL', os.devnull)
+        # x_listening = bool(sp.call('ps au | grep -v grep | grep -i xorg',
+        #                    shell=True, stdout=sp.DEVNULL))
+
+        if self._display is not None:
+            return ':%d' % self._display.vdisplay_num
+
+        sysdisplay = None
+        if self._config.has_option('execution', 'display_variable'):
+            sysdisplay = self._config.get('execution', 'display_variable')
+
+        sysdisplay = sysdisplay or os.getenv('DISPLAY')
+        if sysdisplay:
+            from collections import namedtuple
+
+            def _mock():
+                pass
+
+            # Store a fake Xvfb object
+            ndisp = int(sysdisplay.split(':')[-1])
+            Xvfb = namedtuple('Xvfb', ['vdisplay_num', 'stop'])
+            self._display = Xvfb(ndisp, _mock)
+            return sysdisplay
+        else:
+            if 'darwin' in sys.platform:
+                raise RuntimeError(
+                    'Xvfb requires root permissions to run in OSX. Please '
+                    'make sure that an X server is listening and set the '
+                    'appropriate config on either $DISPLAY or nipype\'s '
+                    '"display_variable" config. Valid X servers include '
+                    'VNC, XQuartz, or manually started Xvfb.')
+
+            # If $DISPLAY is empty, it confuses Xvfb so unset
+            if sysdisplay == '':
+                del os.environ['DISPLAY']
+            try:
+                from xvfbwrapper import Xvfb
+            except ImportError:
+                raise RuntimeError(
+                    'A display server was required, but $DISPLAY is not defined '
+                    'and Xvfb could not be imported.')
+
+            self._display = Xvfb(nolisten='tcp')
+            self._display.start()
+
+            # Older versions of Xvfb used vdisplay_num
+            if hasattr(self._display, 'vdisplay_num'):
+                return ':%d' % self._display.vdisplay_num
+
+            if hasattr(self._display, 'new_display'):
+                return ':%d' % self._display.new_display
+
+    def stop_display(self):
+        """Closes the display if started"""
+        if self._display is not None:
+            self._display.stop()
+
+
+@atexit.register
+def free_display():
+    """Stop virtual display (if it is up)"""
+    from nipype import config
+    from nipype import logging
+    config.stop_display()
+    logging.getLogger('interface').info('Closing display (if virtual)')
```

Comment:
If nipype is going to spawn many `Xvfb`s, shouldn't the stop be associated with the interface, rather than just closing one when Python quits?

I think I must be missing something here.

Reply:
Actually the opposite. You want nipype to run only one `Xvfb` if possible. Let's say you start an ipython session, run an interface with `_redirect_x = True` and then another interface than again requires X. With the current implementation only one `Xvfb` is started. Only when you exit ipython the Xvfb service is killed.

The rationale is precisely stop spawning many `Xvfb`s which was the case before

Reply:
I see. And you're testing to make sure you're not producing multiple by running multiple interfaces? Because my [first comment](https://github.com/nipy/nipype/pull/2203#discussion_r142490521) noted that it looks like each Node makes a copy of `config` before spawning an `xvfb`, so it's not clear they would actually share it.

Again, this is from a very cursory look, so it's possible I'm missing things.

Reply:
@effigies - config makes copies of sections, not the entire config object. so as long as the config.get_display is called, nipype maintains a global config/state.